File: src/main/java/com/heybys/oddments/base/domain/BaseEntity.java
package com.heybys.oddments.base.domain;

import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Transient;

import lombok.Getter;

@Getter
@MappedSuperclass
abstract class BaseEntity implements BaseEntityAware {

    @Column(length = 512, updatable = false)
    private String createdBy;

    @Column(updatable = false)
    private LocalDateTime createdDate;

    @Column(length = 512)
    private String lastModifiedBy;

    @Column
    private LocalDateTime lastModifiedDate;

    @Transient
    private boolean isCreatedByAwareMode = true;

    @Transient
    private boolean isCreatedDateAwareMode = true;

    @Transient
    private boolean isLastModifiedByAwareMode = true;

    @Transient
    private boolean isLastModifiedDateAwareMode = true;

    @PrePersist
    private void prePersist() {
        createdBy = isCreatedByAwareMode ? getCurrentAuditor() : createdBy;
        createdDate = isCreatedDateAwareMode ? getNow() : createdDate;
        lastModifiedBy = isLastModifiedByAwareMode ? getCurrentAuditor() : lastModifiedBy;
        lastModifiedDate = isLastModifiedDateAwareMode ? getNow() : lastModifiedDate;
    }

    @PreUpdate
    private void preUpdate() {
        lastModifiedBy = isLastModifiedByAwareMode ? getCurrentAuditor() : lastModifiedBy;
        lastModifiedDate = isLastModifiedDateAwareMode ? getNow() : lastModifiedDate;
    }

    protected void setCreatedBy(String createdBy) {
        isCreatedByAwareMode = false;
        this.createdBy = createdBy;
    }

    protected void setCreatedDate(LocalDateTime createdDate) {
        isCreatedDateAwareMode = false;
        this.createdDate = createdDate;
    }

    protected void setLastModifiedBy(String lastModifiedBy) {
        isLastModifiedByAwareMode = false;
        this.lastModifiedBy = lastModifiedBy;
    }

    protected void setLastModifiedDate(LocalDateTime lastModifiedDate) {
        isLastModifiedDateAwareMode = false;
        this.lastModifiedDate = lastModifiedDate;
    }
}

File: src/main/kotlin/com/heybys/oddments/base/domain/BaseEntity.kt
package com.heybys.oddments.base.domain

import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import jakarta.persistence.Transient
import java.time.LocalDateTime

@MappedSuperclass
abstract class BaseEntity : BaseEntityAware {

    @Column(length = 512, updatable = false)
    protected var createdBy: String? = null
        private set

    @Column(updatable = false)
    protected var createdDate: LocalDateTime? = null
        private set

    @Column(length = 512)
    protected var lastModifiedBy: String? = null
        private set

    @Column
    protected var lastModifiedDate: LocalDateTime? = null
        private set

    @Transient private var isCreatedByAwareMode = true

    @Transient private var isCreatedDateAwareMode = true

    @Transient private var isLastModifiedByAwareMode = true

    @Transient private var isLastModifiedDateAwareMode = true

    @PrePersist
    private fun prePersist() {
        createdBy = if (isCreatedByAwareMode) currentAuditor else createdBy
        createdDate = if (isCreatedDateAwareMode) now else createdDate
        lastModifiedBy = if (isLastModifiedByAwareMode) currentAuditor else lastModifiedBy
        lastModifiedDate = if (isLastModifiedDateAwareMode) now else lastModifiedDate
    }

    @PreUpdate
    private fun preUpdate() {
        lastModifiedBy = if (isLastModifiedByAwareMode) currentAuditor else lastModifiedBy
        lastModifiedDate = if (isLastModifiedDateAwareMode) now else lastModifiedDate
    }

    protected fun setCreatedBy(createdBy: String) {
        isCreatedByAwareMode = false
        this.createdBy = createdBy
    }

    protected fun setCreatedDate(createdDate: LocalDateTime) {
        isCreatedDateAwareMode = false
        this.createdDate = createdDate
    }

    protected fun setLastModifiedBy(lastModifiedBy: String) {
        isLastModifiedByAwareMode = false
        this.lastModifiedBy = lastModifiedBy
    }

    protected fun setLastModifiedDate(lastModifiedDate: LocalDateTime) {
        isLastModifiedDateAwareMode = false
        this.lastModifiedDate = lastModifiedDate
    }

    // Getter methods for access
    fun getCreatedByValue(): String? = createdBy

    fun getCreatedDateValue(): LocalDateTime? = createdDate

    fun getLastModifiedByValue(): String? = lastModifiedBy

    fun getLastModifiedDateValue(): LocalDateTime? = lastModifiedDate
}
